src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,257 @@
+
+        printf("%s [OPTIONS...]\n\n"
+               "Set up iocost model and qos solutions for block devices\n"
+               "\nCommands:\n"
+               "  apply <path> [solution]    Apply the known solution for the device, if any, otherwise does nothing\n"
+               "  query <path>               Query the known solution for the device\n"
+               "\nOptions:\n"
+               "  -h --help                  Show this help\n"
+               "     --version               Show package version\n"
+               "\nSee the %s for details.\n",
+               program_invocation_short_name,
+               link);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+static int get_known_solutions(sd_device *device, char ***solutions) {
+        const char *value;
+        int r;
+
+        r = sd_device_get_property_value(device, "IOCOST_SOLUTIONS", &value);
+        if (r < 0) {
+                return 1;
+        }
+
+        *solutions = strv_split(value, " ");
+        if (!*solutions)
+                return log_oom();
+
+        return 0;
+}
+
+static int query_named_solution(sd_device *device, const char *name, const char **model, const char **qos) {
+        _cleanup_free_ char *upper_name = NULL, *qos_key = NULL, *model_key = NULL;
+        _cleanup_strv_free_ char **solutions = NULL;
+        int r;
+
+        /* If NULL is passed we query the default solution, which is the first one listed
+         * in the SOLUTIONS key.
+         */
+        if (name == NULL) {
+                r = get_known_solutions(device, &solutions);
+
+                /* A positive 1 return indicates no solutions exist for the device, which should not
+                 * be treated as an error.
+                 */
+                if (r < 0)
+                        return r;
+                else if (r == 1)
+                        return r;
+
+                name = ASSERT_PTR(solutions[0]);
+        }
+
+        upper_name = strdup(name);
+        if (!upper_name)
+                return log_oom();
+
+        ascii_strupper(upper_name);
+
+        if (asprintf(&qos_key, "IOCOST_QOS_%s", upper_name) < 0)
```

Comment:
when just concatenating strings, plesae do not use asprintf() but use strjoin(). it's a million times faster